gitea doctor: panic: reflect: NumField of non-struct type

<!-- NOTE: If your issue is a security concern, please send an email to security@gitea.io instead of opening a public issue -->

<!--
    1. Please speak English, this is the language all maintainers can speak and write.
    2. Please ask questions or configuration/deploy problems on our Discord 
       server (https://discord.gg/gitea) or forum (https://discourse.gitea.io).
    3. Please take a moment to check that your issue doesn't already exist.
    4. Please give all relevant information below for bug reports, because 
       incomplete details will be handled as an invalid report.
-->

- Gitea version (or commit ref): 1.11.6
- Git version: 2.20.1
- Operating system: Linux Debian 10.4 (Buster)
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist: [doctor_1.11.6.log](https://gist.github.com/b-ehlers/38c765b7a26c4c1bdb2bab500cf81004)

## Description

Running gitea doctor results in a panic, see below. The doctor.log is attached as a gist.

I haven't noticed any issues running gitea.

```
git@netcup:~/gitea$ ./gitea doctor all
[ 1 ] Check paths and basic configuration
- Configuration File Path:   '/home/git/gitea/custom/conf/app.ini'
- Repository Root Path:      '/home/git/gitea-repositories'
- Data Root Path:            '/home/git/gitea/data'
- Custom File Root Path:     '/home/git/gitea/custom'
- Work directory:            '/home/git/gitea'
- Log Root Path:             '/home/git/gitea/log'
OK.

[ 2 ] Check Database Version
OK.

[ 3 ] Check if OpenSSH authorized_keys file is up-to-date
OK.

[ 4 ] Check consistency of database
panic: reflect: NumField of non-struct type

goroutine 1 [running]:
reflect.(*rtype).NumField(0x1b6ea20, 0x1b6ea20)
	/usr/local/go/src/reflect/type.go:981 +0x5e
code.gitea.io/gitea/vendor/xorm.io/xorm.(*Engine).mapType(0xc000144b60, 0x1b6ea20, 0x32e2b40, 0x98, 0x1b6ea00, 0x8, 0x18)
	/go/src/code.gitea.io/gitea/vendor/xorm.io/xorm/engine.go:929 +0x2a4
code.gitea.io/gitea/vendor/xorm.io/xorm.(*Engine).autoMapType(0xc000144b60, 0x1b6ea20, 0x32e2b40, 0x98, 0x0, 0x0, 0x0)
	/go/src/code.gitea.io/gitea/vendor/xorm.io/xorm/engine.go:855 +0x193
code.gitea.io/gitea/vendor/xorm.io/xorm.(*Statement).setRefBean(0xc001a8ee18, 0x1b6ea20, 0x32e2b40, 0x3349960, 0x0)
	/go/src/code.gitea.io/gitea/vendor/xorm.io/xorm/statement.go:222 +0x69
code.gitea.io/gitea/vendor/xorm.io/xorm.(*Statement).genCountSQL(0xc001a8ee18, 0xc000b832d0, 0x1, 0x1, 0x1e978a7, 0xc, 0xc000b83190, 0xb93b59, 0xc001ba51a0, 0x1b6ea20, ...)
	/go/src/code.gitea.io/gitea/vendor/xorm.io/xorm/statement.go:992 +0x3f0
code.gitea.io/gitea/vendor/xorm.io/xorm.(*Session).Count(0xc001a8ee00, 0xc000b832d0, 0x1, 0x1, 0x0, 0x0, 0x0)
	/go/src/code.gitea.io/gitea/vendor/xorm.io/xorm/session_stats.go:24 +0x259
code.gitea.io/gitea/cmd.runDoctorCheckDBConsistency(0xc0000c7340, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/code.gitea.io/gitea/cmd/doctor.go:521 +0x26b
code.gitea.io/gitea/cmd.runDoctor(0xc0000c7340, 0x0, 0xc0000c71e0)
	/go/src/code.gitea.io/gitea/cmd/doctor.go:210 +0x9ca
code.gitea.io/gitea/vendor/github.com/urfave/cli.HandleAction(0x1bc5ee0, 0x2ea6b88, 0xc0000c7340, 0xc0003bd700, 0x0)
	/go/src/code.gitea.io/gitea/vendor/github.com/urfave/cli/app.go:490 +0xc8
code.gitea.io/gitea/vendor/github.com/urfave/cli.Command.Run(0x1e85db0, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1ea4b37, 0x11, 0x0, ...)
	/go/src/code.gitea.io/gitea/vendor/github.com/urfave/cli/command.go:210 +0x991
code.gitea.io/gitea/vendor/github.com/urfave/cli.(*App).Run(0xc00009b1e0, 0xc00003c180, 0x3, 0x3, 0x0, 0x0)
	/go/src/code.gitea.io/gitea/vendor/github.com/urfave/cli/app.go:255 +0x6ab
main.main()
	/go/src/code.gitea.io/gitea/main.go:110 +0x845
git@netcup:~/gitea$
```

--- a/cmd/doctor.go
+++ b/cmd/doctor.go
@@ -502,33 +502,21 @@ func runDoctorScriptType(ctx *cli.Context) ([]string, error) {
 }
 
 func runDoctorCheckDBConsistency(ctx *cli.Context) ([]string, error) {
+	var results []string
+
 	// make sure DB version is uptodate
 	if err := models.NewEngine(context.Background(), migrations.EnsureUpToDate); err != nil {
 		return nil, fmt.Errorf("model version on the database does not match the current Gitea version. Model consistency will not be checked until the database is upgraded")
 	}
-	_, committer, err := models.TxDBContext()
-	if err != nil {
-		return nil, err
-	}
-	sess := committer.(models.Engine)
-	defer committer.Close()
-	var results []string
 
 	//find tracked times without existing issues/pulls
-	count, err := sess.Table("tracked_time").
-		Join("LEFT", "issue", "tracked_time.issue_id=issue.id").
-		Where("issue.id is NULL").
-		Count("id")
+	count, err := models.CountOrphanedObjects("tracked_time", "issue", "tracked_time.issue_id=issue.id")
 	if err != nil {
 		return nil, err
 	}
 	if count > 0 {
 		if ctx.Bool("fix") {
-			if _, err = sess.In("id", builder.Select("tracked_time.id").
-				From("tracked_time").
-				Join("LEFT", "issue", "tracked_time.issue_id=issue.id").
-				Where(builder.IsNull{"issue.id"})).
-				Delete(models.TrackedTime{}); err != nil {
+			if err = models.DeleteOrphanedObjects("tracked_time", "issue", "tracked_time.issue_id=issue.id"); err != nil {
 				return nil, err
 			}
 			results = append(results, fmt.Sprintf("%d tracked times without existing issue deleted", count))
@@ -537,8 +525,5 @@ func runDoctorCheckDBConsistency(ctx *cli.Context) ([]string, error) {
 		}
 	}
 
-	if ctx.Bool("fix") {
-		return results, committer.Commit()
-	}
 	return results, nil
 }

--- a/models/consistency.go
+++ b/models/consistency.go
@@ -171,10 +171,12 @@ func (action *Action) checkForConsistency(t *testing.T) {
 
 // CountOrphanedObjects count subjects with have no existing refobject anymore
 func CountOrphanedObjects(subject, refobject, joinCond string) (int64, error) {
-	return x.Table("`"+subject+"`").
+	var ids []int64
+
+	return int64(len(ids)), x.Table("`"+subject+"`").
 		Join("LEFT", refobject, joinCond).
 		Where(builder.IsNull{"`" + refobject + "`.id"}).
-		Count("id")
+		Select("id").Find(&ids)
 }
 
 // DeleteOrphanedObjects delete subjects with have no existing refobject anymore

--- a/models/consistency.go
+++ b/models/consistency.go
@@ -10,6 +10,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"xorm.io/builder"
 )
 
 // consistencyCheckable a type that can be tested for database consistency
@@ -166,4 +167,22 @@ func (team *Team) checkForConsistency(t *testing.T) {
 func (action *Action) checkForConsistency(t *testing.T) {
 	repo := AssertExistsAndLoadBean(t, &Repository{ID: action.RepoID}).(*Repository)
 	assert.Equal(t, repo.IsPrivate, action.IsPrivate, "action: %+v", action)
+}
+
+// CountOrphanedObjects count subjects with have no existing refobject anymore
+func CountOrphanedObjects(subject, refobject, joinCond string) (int64, error) {
+	return x.Table("`"+subject+"`").
+		Join("LEFT", refobject, joinCond).
+		Where(builder.IsNull{"`" + refobject + "`.id"}).
+		Count("id")
+}
+
+// DeleteOrphanedObjects delete subjects with have no existing refobject anymore
+func DeleteOrphanedObjects(subject, refobject, joinCond string) error {
+	_, err := x.In("id", builder.Select("`"+subject+"`.id").
+		From("`"+subject+"`").
+		Join("LEFT", "`"+refobject+"`", joinCond).
+		Where(builder.IsNull{"`" + refobject + "`.id"})).
+		Delete("`" + subject + "`")
+	return err
 }

